include/mbedtls/platform_util.h

```diff
@@ -167,7 +167,28 @@ MBEDTLS_DEPRECATED typedef int mbedtls_deprecated_numeric_constant_t;
  * \param len   Length of the buffer in bytes
  *
  */
+#if defined(MBEDTLS_PLATFORM_ZEROIZE_CHECK_UNSAFE)
+#define MBEDTLS_PLATFORM_ZEROIZE_ALT
+#define mbedtls_platform_zeroize(buf, len) memset(buf, 0, len)
+#include <string.h>
+#else
 void mbedtls_platform_zeroize(void *buf, size_t len);
+#endif
+
+/* MBEDTLS_PLATFORM_ZEROIZE_CHECK_UNSAFE
+ *
+ * Replaces calls to mbedtls_platform_zeroize() with calls to memset(),
+ * to allow compiler analysis to check for invalid length arguments (e.g.
+ * specifying sizeof(pointer) rather than sizeof(pointee)).
+ *
+ * Note that this option is meant for internal use only and must not be used
+ * in production builds, because that would lead to zeroization calls being
+ * optimised out by the compiler.
+ *
+ * It is only intended to be used in CFLAGS, with -Wsizeof-pointer-memaccess,
+ * to check for those incorrect calls to mbedtls_platform_zeroize().
+ */
+//#define MBEDTLS_PLATFORM_ZEROIZE_CHECK_UNSAFE
```

Comment:
I really don't like having test-only code that reduces security in the library. Our [normal policy](https://github.com/Mbed-TLS/mbedtls/blob/development/docs/architecture/testing/invasive-testing.md) for this is a two-tier system: enable a compilation option that by itself doesn't change the runtime behavior, and provide or run additional code that has to be deliberately written.

Here, we can already achieve the effect by activating `MBEDTLS_PLATFORM_ZEROIZE_ALT` and building with an implementation of `mbedtls_platform_zeroize`, which lives in the `tests/src` directory, that just calls `memset`. (Similar example: we can build with a deterministic, non-cryptographic RNG by activating `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` and linking with the external RNG from `tests/src/fake_external_rng_for_test.c`.) See [`tests/configs/user-config-for-test.h`](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/configs/user-config-for-test.h#L61) and [`tests/configs/config-wrapper-malloc-0-null.h`](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/configs/config-wrapper-malloc-0-null.h) for examples doing something similar.

Reply:
We actually need the compiler to see the calls to `mbedtls_platform_zeroize()` as `memset()`, rather to have a call to `mbedtls_platform_zeroize()` be implemented as `memset()`, since it's at the `mbedtls_platform_zeroize()` call site that the information is available about the types of parameters, which is what need to be checked.

This can only be done with a macro, so the choice is between having it in the header permanently, like this, or having the test use a `perl` (or `sed`, I suppose) script to make these changes. However, that would be fragile.

Also, I disagree that this is test-only code: rather, it is enabling static analysis. If there was an annotation comparable to the `__attribute__ ((format (...)))` one for `printf`-style function then we would apply it to this function definition at this point; this is very similar to that, so I think this is the right way to do this.

Reply:
See comment below - we can add a check for `MBEDTLS_TEST_HOOKS` to limit the risk?

Reply:
It's annoying that this can't be done with the existing option. (I checked, and indeed even compiling with `-O3` and a `static inline` wrapper with `__attribute__((access(write_only, 1, 2)))` is not enough to get the warning.) Food for thought when we [simplify the platform interface](https://github.com/Mbed-TLS/mbedtls/issues/8108): this is a thing you can't do with a function replacement, only if you 
can define a macro, the blocking thing being the function prototype. (If we manage to skip the function prototype then we can define a macro and all is good.)

So this does require a patch to `platform_util.h`. But (unless we actually patch the file before building, which I'd rather avoid), this patch should still not, by itself, make an insecure build as easy as `-D`. Gating by `MBEDTLS_TEST_HOOKS` doesn't help: there's already a compile-time option to gate the insecurity (which should be called `MBEDTLS_TEST_something`), but our policy is a two-tier system to make it a genuine hurdle to make an insecure build: an insecure build should need to either write your own insecure code or compile with something under `tests`.

The minimum patch we need in `platform_util.h` is to skip the prototype. So I propose to have a symbol (not referenced outside `platform_util.h` and test code) that does just that. The `#define mbedtls_platform_zeroize …` should be in a file under `tests`.